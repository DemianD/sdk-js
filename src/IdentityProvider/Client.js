import filter from '../../filter'
import request from '../request'
import retry from '../../retry'

const MAX_ATTEMPTS = 3
const REFETCH_WINDOW = 20
const S_TO_MS = 1000

export default class ClientIdentityProvider {
  constructor(options = {}) {
    this._options = filter(options, 'Client', 'Realm')

    if (!this._options.Client) {
      throw new TypeError(
        'Client is a required option for ClientIdentityProvider'
      )
    }
    if (!this._options.Realm) {
      throw new TypeError(
        'Realm is a required option for ClientIdentityProvider'
      )
    }
  }

  getAuthorization(force = false) {
    if (this._current === undefined || force) {
      this._current = retry(
        (resolve, reject) => {
          request(
            `${this._options.Realm}/access_tokens`,
            {
              grant_type: 'public_client',
              client_id: this._options.Client,
            },
            {
              method: 'POST',
            }
          ).then(resolve, reject)
        },
        {
          max_attempts: MAX_ATTEMPTS,
        }
      )
        .then(access_token_response => ({
          Authorization: `${access_token_response.body.token_type} ${
            access_token_response.body.access_token
          }`,
          Expiration: access_token_response.body.expires_in,
          Realm: this._options.Realm,
        }))
        .then(access => {
          setTimeout(
            this.getAuthorization.bind(this, true),
            (access.Expiration - REFETCH_WINDOW) * S_TO_MS
          )
          return access
        })
    }
    return this._current
  }
}
