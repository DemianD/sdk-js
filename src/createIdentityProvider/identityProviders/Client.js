import network from '../../util/network'

import pick from '../../util/pick'
import retry from '../../util/retry'

import { createIdentityProviderError } from '../../util/createError'

import {
  MISCONFIGURED,
  MISCONFIGURED_CLIENT_ID,
  MISCONFIGURED_REALM,
} from '../../constants'

const MAX_ATTEMPTS = 3
const REFETCH_WINDOW = 20
const S_TO_MS = 1000

export default class Client {
  constructor(options = {}) {
    this._options = pick(options, 'client_id', 'realm')

    if (!this._options.client_id) {
      throw createIdentityProviderError(
        'client_id is a required option for `Client`',
        MISCONFIGURED,
        MISCONFIGURED_CLIENT_ID
      )
    }

    if (!this._options.realm) {
      throw createIdentityProviderError(
        'realm is a required option for `Client`',
        MISCONFIGURED,
        MISCONFIGURED_REALM
      )
    }
  }

  getAuthorization(force = false) {
    if (this._current === undefined || force) {
      this._current = retry(
        (resolve, reject) => {
          network
            .post(`${this._options.realm}/access_tokens`, {
              grant_type: 'public_client',
              client_id: this._options.client_id,
            })
            .then(resolve, reject)
        },
        {
          max_attempts: MAX_ATTEMPTS,
        }
      )
        .then(({ data }) => ({
          Authorization: `${data.token_type} ${data.access_token}`,
          Expiration: data.expires_in,
          Realm: this._options.realm,
        }))
        .then(({ Realm, Authorization, Expiration }) =>
          network
            .get(`${Realm}/integrations/proxy`, {
              headers: { Authorization },
            })
            .then(({ data: response }) => ({
              Realm: response.data.url,
              Authorization,
              Expiration,
            }))
        )
        .then(access => {
          setTimeout(
            this.getAuthorization.bind(this, true),
            (access.Expiration - REFETCH_WINDOW) * S_TO_MS
          )
          return access
        })
    }

    return this._current
  }
}
