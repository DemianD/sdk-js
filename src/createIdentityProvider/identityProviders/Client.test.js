/* eslint-disable max-nested-callbacks */
import moxios from 'moxios'
import axios from 'axios'

import Client from './Client'

describe('Identity Provider # Client', () => {
  let network
  beforeEach(() => {
    network = axios.create()
    moxios.install(network)
  })

  afterEach(() => {
    moxios.uninstall(network)
  })

  it('should be able to create an identity provider instance', () => {
    const client = new Client(network, {
      client_id: 'client 123',
      realm: 'https://example.com',
    })

    expect(client).toBeDefined()
  })

  it('should error when no options are given', () => {
    expect(() => new Client(network)).toThrowErrorMatchingSnapshot()
  })

  it('should error when client_id is not given', () => {
    expect(
      () =>
        new Client(network, {
          realm: 'https://example.com',
        })
    ).toThrowErrorMatchingSnapshot()
  })

  it('should error when realm is not given', () => {
    expect(
      () =>
        new Client(network, {
          client_id: 'client 123',
        })
    ).toThrowErrorMatchingSnapshot()
  })

  it('should be possible to authenticate', async () => {
    const client = new Client(network, {
      client_id: 'client 123',
      realm: 'https://example.com',
    })

    // Mock the authentication request
    moxios.wait(() => {
      moxios.requests.mostRecent().respondWith({
        status: 200,
        response: {
          access_token: 'rMOUIcH85oh44KD7RM4XRk7jFPiG8RMLi2IPqFsQ',
          token_type: 'Bearer',
          expires_in: 5400,
        },
      })

      // Mock the proxy call
      moxios.wait(() => {
        moxios.requests.mostRecent().respondWith({
          status: 200,
          response: {
            data: {
              url: 'https://api.example.com',
            },
          },
        })
      }, 0)
    }, 0)

    // Wait for our call
    await client.getAuthorization()

    const {
      data,
      headers,
      method,
      params,
      url,
    } = moxios.requests.mostRecent().config

    expect({ data, headers, method, params, url }).toMatchSnapshot()

    // A call to auth + proxy should have been made
    expect(moxios.requests.count()).toBe(2)
  })

  it('should be possible to authenticate with a resource_code', async () => {
    const client = new Client(network, {
      client_id: 'client 123',
      realm: 'https://example.com',
      resource_code: 'some resource code',
    })

    // Mock the authentication request
    moxios.wait(() => {
      moxios.requests.mostRecent().respondWith({
        status: 200,
        response: {
          access_token: 'rMOUIcH85oh44KD7RM4XRk7jFPiG8RMLi2IPqFsQ',
          token_type: 'Bearer',
          expires_in: 5400,
        },
      })

      // Mock the proxy call
      moxios.wait(() => {
        moxios.requests.mostRecent().respondWith({
          status: 200,
          response: {
            data: {
              url: 'https://api.example.com',
            },
          },
        })
      }, 0)
    }, 0)

    // Wait for our call
    await client.getAuthorization()

    const {
      data,
      headers,
      method,
      params,
      url,
    } = moxios.requests.mostRecent().config

    expect({ data, headers, method, params, url }).toMatchSnapshot()

    // A call to auth + proxy should have been made
    expect(moxios.requests.count()).toBe(2)
  })

  it('should be possible to re-use the same auth request', async () => {
    const client = new Client(network, {
      client_id: 'client 123',
      realm: 'https://example.com',
    })

    // Mock the authentication request
    moxios.wait(() => {
      moxios.requests.mostRecent().respondWith({
        status: 200,
        response: {
          access_token: 'rMOUIcH85oh44KD7RM4XRk7jFPiG8RMLi2IPqFsQ',
          token_type: 'Bearer',
          expires_in: 5400,
        },
      })

      // Mock the proxy call
      moxios.wait(() => {
        moxios.requests.mostRecent().respondWith({
          status: 200,
          response: {
            data: {
              url: 'https://api.example.com',
            },
          },
        })
      }, 0)
    }, 0)

    // Wait for our call
    await client.getAuthorization()
    await client.getAuthorization()
    await client.getAuthorization()

    // Only the first auth + proxy call should have been made
    expect(moxios.requests.count()).toBe(2)
  })
})
