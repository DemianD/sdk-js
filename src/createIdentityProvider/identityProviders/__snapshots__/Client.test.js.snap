// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Identity Provider # Client should be possible to authenticate 1`] = `
Object {
  "data": undefined,
  "headers": Object {
    "Accept": "application/json, text/plain, */*",
    "Authorization": "Bearer rMOUIcH85oh44KD7RM4XRk7jFPiG8RMLi2IPqFsQ",
  },
  "method": "get",
  "params": undefined,
  "url": "https://example.com/integrations/proxy",
}
`;

exports[`Identity Provider # Client should error when client_id is not given 1`] = `"[IDENTITY PROVIDER]: client_id is a required option for \`Client\`"`;

exports[`Identity Provider # Client should error when no options are given 1`] = `"[IDENTITY PROVIDER]: client_id is a required option for \`Client\`"`;

exports[`Identity Provider # Client should error when realm is not given 1`] = `"[IDENTITY PROVIDER]: realm is a required option for \`Client\`"`;
